### PR TITLE
8273916: Remove 'special' ranking

### PR DESCRIPTION
**Pull request by coleenp**
This change removes the special ranking and folds it into nosafepoint.  You have to look at commit #3 to see this actual part of the change that doesn't include JDK-8273915.
This passes tier1-6 also.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8273916](https://bugs.openjdk.java.net/browse/JDK-8273916): Remove 'special' ranking

### Reviewers
 * [David Holmes](https://openjdk.java.net/census#dholmes) (@dholmes-ora - **Reviewer**)
 * [Patricio Chilano Mateo](https://openjdk.java.net/census#pchilanomate) (@pchilano - Committer) ⚠️ Review applies to 5867b9a55cb4c4068a87535fb614127ec89fe7ee

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/5563/head:pull/5563` \
`$ git checkout pull/5563`

Update a local copy of the PR: \
`$ git checkout pull/5563` \
`$ git pull https://git.openjdk.java.net/jdk pull/5563/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 5563`

View PR using the GUI difftool: \
`$ git pr show -t 5563`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/5563.diff">https://git.openjdk.java.net/jdk/pull/5563.diff</a>

</details>

### CODE REVIEW

**Review by dholmes-ora — COMMENTED** [review]
Sorry Coleen but I'm not understanding the mapping process here. I expected to see all special changed to the same thing, eg. nosafepoint, , and all special-N changed to nosafepoint-N, but you have not done that. ???

David

**Comment by coleenp** [comment]
Thanks for looking at this, David. I started to map special(-n) to nosafepoint(-n) but since special is nosafepoint - 9 (?) there were interactions with other nosafepoint locks so they needed rankings relative to other locks.

CompiledMethod_lock(nosafepoint-4) -> CodeCache_lock(nosafepoint-3) -> VtableStubs_lock(nosafepoint-2) -> CompiledIC_lock(nosafepoint)

CodeSweeper_lock(nosafepoint-5) -> CompiledMethod_lock(nosafepoint-4)

ThreadsSMRDelete_lock(nosafepoint-2) ->  (can't remember which one was nosafepoint-1 anymore!)

The compiler locks have the deepest nestings.

**Comment by dholmes-ora** [comment]
Sorry I don't follow. Lets examine that first example. We currently have:

CompiledMethod_lock (special-1) ->
CodeCache_lock (special) ->
VtableStubs_lock (leaf-2 == special + 11) ->
CompiledIC_lock(leaf+2 == special + 15)

If we change special to nosafepoint then we would have:

CompiledMethod_lock (nosafepoint-1) ->
CodeCache_lock (nosafepoint) ->
VtableStubs_lock (leaf-2 == nosafepoint + 4) ->
CompiledIC_lock(leaf+2 == nosafepoint + 8 == safepoint - 8)

so I don't see why we would have CodeCache_lock instead be nosafepoint-3 ?

If all you have done is change special to nosafepoint then all the existing relative rankings remain for things like leaf and leaf-2. The only adjustment you have to make is for a leaf+2 to instead be expressed as safepoint-8.

David

**Comment by coleenp** [comment]
All the lock rankings are negative values from locks they hold.  The nosafepoint ranking and ranking below that are safepoint_check_never locks.  So nosafepoint is the top of the lock hierarchy for safepoint_check_never locks.  That's how we start with CompiledIC_lock at 'nosafepoint' rank.

The lock values will not be exactly the same value as they were when they were 'special'.  They'll be ordered relative to their ranking in the nosafepoint to nosafepoint-n range.

I did skip a value though (nosafepoint-1).

**Comment by dholmes-ora** [comment]
So there is more to this than just removing the "special" ranking - you've also changed some locks that are safepoint_never, that used to have ranks above what is now nosafepoint, so that they instead have ranks below nosafepoint - is that right? As long a all relative rankings of locks that can be taken together is maintained, then that is okay - but it is very hard to see that just by looking at the changes.

**Comment by coleenp** [comment]
I just merged with my previous commit and hopefully this change makes a lot more sense now.

File: src/hotspot/share/runtime/mutexLocker.cpp [hunk 1]
```diff
@@ -225,17 +225,17 @@ void mutex_init() {
   def(StringDedup_lock             , PaddedMonitor, nosafepoint, true,  _safepoint_check_never);
   def(StringDedupIntern_lock       , PaddedMutex  , nosafepoint, true,  _safepoint_check_never);
   def(ParGCRareEvent_lock          , PaddedMutex  , leaf,        true,  _safepoint_check_always);
-  def(CodeCache_lock               , PaddedMonitor, special,     true,  _safepoint_check_never);
-  def(CodeSweeper_lock             , PaddedMonitor, special-2,   true,  _safepoint_check_never);
-  def(RawMonitor_lock              , PaddedMutex  , special,     true,  _safepoint_check_never);
+  def(CodeCache_lock               , PaddedMonitor, nosafepoint-3, true,  _safepoint_check_never);
```
[the remark this answers is not in the document]
Comment (coleenp) [reply src/hotspot/share/runtime/mutexLocker.cpp:228]: Thanks for noticing that.

**Comment by coleenp** [comment]
Thanks Patricio for the code review!

**Review by dholmes-ora — APPROVED** [review]
Hi Coleen,

Thanks again for the offlist discussion. Sorry it took me so long to "get it".

Cheers,
David

**Comment by coleenp** [comment]
Thank you @dholmes-ora for the discussions and review!
/integrate

**Pull request closed without merging**